Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp

```diff
@@ -2450,9 +2450,25 @@ static void collect_attribute_values_of_an_inheritance_stack(SourceGenerator& fu
 
             if (attribute.extended_attributes.contains("Reflect")) {
                 if (attribute.type->name() != "boolean") {
-                    attribute_generator.append(R"~~~(
+                    if (interface_in_chain.extended_attributes.contains("UseDeprecatedAKString")) {
+                        attribute_generator.append(R"~~~(
+    auto @attribute.return_value_name@ = impl->deprecated_attribute(HTML::AttributeNames::@attribute.reflect_name@);
+)~~~");
+                    } else {
+                        if (attribute.type->is_nullable()) {
+                            attribute_generator.append(R"~~~(
     auto @attribute.return_value_name@ = impl->attribute(HTML::AttributeNames::@attribute.reflect_name@);
 )~~~");
+                        } else if (attribute.extended_attributes.contains("LegacyNullToEmptyString")) {
+                            attribute_generator.append(R"~~~(
+    auto @attribute.return_value_name@ = impl->attribute(HTML::AttributeNames::@attribute.reflect_name@).value_or(String {});
+)~~~");
+                        } else {
+                            attribute_generator.append(R"~~~(
+    auto @attribute.return_value_name@ = impl->attribute(HTML::AttributeNames::@attribute.reflect_name@).value_or("null"_string);
```

Comment:
The main change which I am not 100% sure about yet is this one. I _think_ this is what is meant to be done here from reading the spec, but am not certain on that at all.

Reply:
I would expect it to return a JS `null`, not the string `"null"`
This seems like something you could check with other browsers. IIUC nullity here just means that the attribute is not present at all on the element.

Reply:
hmmm I've just been testing some more, and it seems that other browsers may just return the empty string?

```html
<!DOCTYPE html>
<html>
<body>
<script id="thing"></script>
<script>
console.log(document.getElementById("thing").type);
</script>
</body>
</html>
```
> (empty string)

Seems like I'm breaking that behaviour with this change, I'll add a test for it and will fix. I'm now quite confused as to what `LegacyNullToEmptyString` means then :thinking:

Part of my confusion around this stems around an example like `DOMException`  or `CharacterData`. They accept a `optional DOMString message = ""` in the constructor. You can pass `null` through, and it seems it will just get sringified to the string "null":

```js
let e = new DOMException(null, null);
console.log(e.message);
```
> "null"

I'll try read the spec some more, I'm definitely misunderstanding something :thinking:

Reply:
Maybe the correct change is to only do:

 ```cpp
    auto ret = impl->attribute(HTML::AttributeNames::@attribute.reflect_name@).value_or(String {});
```

here

Reply:
See the spec for reflect here: https://dom.spec.whatwg.org/#concept-reflect

For getting the attribute, we're meant to use: https://dom.spec.whatwg.org/#concept-element-attributes-get-value
And for setting the attribute, we're meant to use: https://dom.spec.whatwg.org/#concept-element-attributes-set-value

These are `Element::get_attribute_value`  and `Element::set_attribute_value`, respectively, which were recently added in #20892

Reply:
Thanks! I was looking at the wrong spec for this!

And nice, nothing calls them yet, so should be easy to port over and include here :^)

Reply:
Using `get_attribute_value` doees keep the same behavior as before.

In `master`, we currently use `Element::attribute`, which returns a null `DeprecatedString` if the attribute doesn't exist. When a null `DeprecatedString` is converted to a `JS::PrimitiveString`, it becomes the empty string.

Using `get_attribute_value` is the same as `get_attribute`, except it would return an empty string already instead of null.

Reply:
> In master, we currently use Element::attribute, which returns a null DeprecatedString if the attribute doesn't exist. When a null DeprecatedString is converted to a JS::PrimitiveString, it becomes the empty string.

I think using `get_attribute_value` will change behaviour for nullable Reflected return values if I change this to unconditonally call `get_attribute_value` here, unless you're meaning I only call this for the non-nullable type?

>     [Reflect=crossorigin, CEReactions] attribute DOMString? crossOrigin;

Which is generated to:
```cpp
    auto retval = impl->attribute(HTML::AttributeNames::crossorigin);
    // ...
    if (retval.is_null()) {
        return JS::js_null();
    } else {

    return JS::PrimitiveString::create(vm, retval);

    }
```

After switching to new String, it becomes:
```cpp
    auto retval = impl->attribute(HTML::AttributeNames::crossorigin);
    // ...
    if (!retval.has_value()) {
        return JS::js_null();
    } else {

    return JS::PrimitiveString::create(vm, retval.release_value());

    }
```

Are you suggesting I change this for the case that won't be nullable, or for both cases here? That's why I thought my current change is preserving behaviour, and I left it as FIXME since I wasn't entirely sure on what the spec meant for this

Reply:
Ahhh I see, okay let's leave the FIXME for now then.

Looks like nullable Reflect attributes link to https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#reflect
which is a different reflect algorithm than what I linked previously (https://dom.spec.whatwg.org/#concept-reflect)

So there's probably some difference we're supposed to detect, and change which reflect algorithm we use in the generated IDL? In any case, yeah the FIXME is fine for now :sweat_smile:

Reply:
I'll try come back and fix this FIXME later :)